packages/playwright-test/src/reporters/azure.ts

```diff
@@ -0,0 +1,475 @@
+        });
+      }
+      return pointsIds;
+    } catch (e) {
+      this.log(colors.red(e.message));
+    }
+    return pointsIds;
+  }
+
+  private addReportingOverride = (api: Test.ITestApi): Test.ITestApi => {
+    // https://github.com/microsoft/azure-devops-node-api/issues/318#issuecomment-498802402
+    api.addTestResultsToTestRun = function(results, projectName, runId) {
+      return new Promise(async (resolve, reject) => {
+        const routeValues = {
+          project: projectName,
+          runId
+        };
+
+        try {
+          const verData = await this.vsoClient.getVersioningData(
+              '5.0-preview.5',
+              'Test',
+              '4637d869-3a76-4468-8057-0bb02aa385cf',
+              routeValues
+          );
+          const url = verData.requestUrl;
+          const options = this.createRequestOptions('application/json', verData.apiVersion);
+          const res = await this.rest.create(url as string, results, options);
+          resolve(res as any);
+        } catch (err) {
+          reject(err);
+        }
+      });
+    };
+    return api;
+  };
+
+  private async uploadAttachmentsFunc(testResult: TestResult, caseId: number, testCaseId: string): Promise<string[]> {
+    this.log(colors.gray(`Start upload attachments for test case [${testCaseId}]`));
+    const attachmentsResult: string[] = [];
+    for (const attachment of testResult.attachments) {
+      try {
+        if (this.attachmentsType!.includes((attachment.name as TAttachmentType[number]))) {
+          if (existsSync(attachment.path!)) {
+            const attachments: TestInterfaces.TestAttachmentRequestModel = {
+              attachmentType: 'GeneralAttachment',
+              fileName: `${attachment.name}-${createGuid()}.${attachment.contentType.split('/')[1]}`,
+              stream: readFileSync(attachment.path!, { encoding: 'base64' })
+            };
+
+            if (!this.testApi)
+              this.testApi = await this.connection.getTestApi();
+            const response = await this.testApi.createTestResultAttachment(
+                attachments,
+                this.projectName,
+                this.runId!,
+                caseId
+            );
+            attachmentsResult.push(response.url);
+          } else {
+            throw new Error(`Attachment ${attachment.path} does not exist`);
+          }
+        }
+      } catch (err) {
+        this.log(colors.red(err.message));
+      }
+    }
+    return attachmentsResult;
+  }
+
+  private async awaitForRunId(): Promise<number | void> {
+    // need wait runId variable to be initialised in onBegin() hook
+    if (this.isDisabled)
+      return;
+    try {
+      const timeout = 10_000;
+      const startTime = Date.now();
+      while (this.runId === undefined && Date.now() - startTime < timeout)
```

Comment:
this can be done nicer in the following way.

1. inside constructor do the following:
```js
this.runIdPromise = new Promise((resolve, reject) => {
  this.resolveRunId = resolve;
  this.rejectRunId = reject;
});
```
2. inside `onBegin()` remove that it returns a Promise, since its a sync method. and do something like that:
```js
  onBegin(): void {
    if (this.isDisabled)
      return;
    this.connection = new azdev.WebApi(this.orgUrl, azdev.getPersonalAccessTokenHandler(this.token));

    (async () => {
      this.testApi = await this.connection.getTestApi();
      const run = await this.createRun(this.testRunTitle);
      if (run) {
        this.resolveRunId(run.id)
        this.log(colors.green(`Using run ${this.runId} to publish test results`));
      } else {
        this.isDisabled = true;
      }
    })(error => {
      this.isDisabled = true;
      this.warning(`Failed to create test run: ${error.message}. Check your token and orgUrl. Reporting is disabled.`);
      rejectRunId(error);
    });
  }
```
3. in the places you need the run id, just do the following:
```
const runId = await this.runIdPromise;
```